tools/kagami/src/swarm.rs

```diff
@@ -31,166 +31,461 @@ const FILE_VALIDATOR: &str = "validator.wasm";
+            /// Default `config.json`, `genesis.json` and `validator.wasm` are generated and put into
+            /// the `<outdir>/config` directory. That directory is specified in the `volumes` field
+            /// of the Docker Compose file.
+            ///
+            /// Setting this flag prevents copying of default configuration files into the output folder.
+            /// The `config` directory will still be created, but the necessary configuration should be put
+            /// there by the user manually.
+            #[arg(long)]
+            no_default_configuration: bool,
+            #[command(flatten)]
+            source: ModeDirSource,
+        },
+        /// Produce only a single Docker Compose configuration file
+        File {
+            /// Path to a generated Docker Compose configuration.
+            ///
+            /// If file exists, Kagami will prompt its overwriting. If the TTY is not
+            /// interactive, Kagami will stop execution with non-zero exit code. In order to
+            /// overwrite the file anyway, pass `--force` flag.
+            outfile: PathBuf,
+            /// Path to a directory with Iroha configuration. It will be mapped as volume for containers.
+            ///
+            /// The directory should contain `config.json` and `genesis.json`.
+            #[arg(long)]
+            config_dir: PathBuf,
+            #[command(flatten)]
+            source: ModeFileSource,
+        },
+    }
+
+    #[derive(Args, Debug)]
+    #[group(required = true, multiple = false)]
+    pub struct ModeDirSource {
+        /// Use Iroha GitHub source as a build source
+        ///
+        /// Clone `hyperledger/iroha` repo from the revision Kagami is built itself,
+        /// and use the cloned source code to build images from.
+        #[arg(long)]
+        pub build_from_github: bool,
+        /// Use specified docker image.
+        ///
+        /// Be careful with specifying a Dockerhub image as a source: Kagami Swarm only guarantees that
+        /// the docker-compose configuration it generates is compatible with the same Git revision it
+        /// is built from itself. Therefore, if specified image is not compatible with the version of Swarm
+        /// you are running, the generated configuration might not work.
+        #[arg(long)]
+        pub image: Option<String>,
+        /// Use local path location of the Iroha source code to build images from.
+        ///
+        /// If the path is relative, it will be resolved relative to the CWD.
+        #[arg(long, value_name = "PATH")]
+        pub build: Option<PathBuf>,
+    }
+
+    #[derive(Args, Debug)]
+    #[group(required = true, multiple = false)]
+    // FIXME: I haven't found a way how to share `image` and `build` options between `file` and
+    //        `dir` modes with correct grouping logic. `command(flatten)` doesn't work for it,
+    //        so it's hard to share a single struct with "base source options"
```

Comment:
I don't think it's that crucial as it's not too much of reduplication, but maybe trying to make it [global](https://docs.rs/clap/latest/clap/struct.Arg.html#method.global) would work?

Reply:
Resolved in favour of #3647